nntrainer/src/neuralnet.cpp

```diff
@@ -320,18 +314,21 @@ int NeuralNetwork::initLossLayer() {
   int status = ML_ERROR_NONE;
   CostType updated_cost = cost;
 
+  if (layers.empty()) {
+    status = ML_ERROR_INVALID_PARAMETER;
+  }
+
   if (updated_cost == COST_ENTROPY) {
     if (layers.back()->getType() != LAYER_ACTIVATION) {
       ml_loge("Error: Cross Entropy need last layer to have softmax or sigmoid "
               "activation.");
       return ML_ERROR_NOT_SUPPORTED;
     }
 
-    std::shared_ptr<ActivationLayer> act_layer =
-      std::static_pointer_cast<ActivationLayer>(layers.back());
+    std::shared_ptr<Layer> act_layer = layers.back();
     layers.pop_back();
 
-    switch (act_layer->getActivationType()) {
+    switch (layers.back()->getActivationType()) {
```

Comment:
This might lead to nasty bugs. Now, you are getting the activation type from the previous layer which is not activation layer anymore.
Cache this value before popping out the activation layer, and use from it than from the last layer.

Reply:
Yes. Now me is spoiling past me's work lol. will fix that.

Reply:
It is fixed